bindings/cs/MongoDB.Crypt/CryptContext.cs

```diff
@@ -0,0 +1,204 @@
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace MongoDB.Crypt
+{
+    /// <summary>
+    /// A encryption or decryption session. It may not be reused.
+    /// </summary>
+    /// <seealso cref="System.IDisposable" />
+    /// <seealso cref="MongoDB.Crypt.IStatus" />
+    public class CryptContext : IDisposable, IStatus
+    {
+        /// <summary>
+        /// States of the CryptContext state machine
+        /// </summary>
+        public enum StateCode
+        {
+            /// <summary>
+            /// LibMongoCrypt hit an error
+            /// </summary>
+            MONGOCRYPT_CTX_ERROR = 0,
+
+            /// <summary>
+            /// LibMongoCrypt wants the collection information by running a OP_MSG command against the users' mongod.
+            /// </summary>
+            MONGOCRYPT_CTX_NEED_MONGO_COLLINFO = 1,
+
+            /// <summary>
+            /// LibMongoCrypt wants a command to be run against mongocryptd.
+            /// </summary>
+            MONGOCRYPT_CTX_NEED_MONGO_MARKINGS = 2,
+
+            /// <summary>
+            /// LibMongoCrypt wants a command to be run against mongod key vault.
+            /// </summary>
+            MONGOCRYPT_CTX_NEED_MONGO_KEYS = 3,
+
+            /// <summary>
+            /// LibMongoCrypt wants a request sent to KMS.
+            /// </summary>
+            MONGOCRYPT_CTX_NEED_KMS = 4,
+
+            /// <summary>
+            /// LibMongoCrypt is ready to do encryption, call Finish().
+            /// </summary>
+            MONGOCRYPT_CTX_READY = 5,
+
+            /// <summary>
+            /// LibMongoCrypt is complete.
+            /// </summary>
+            MONGOCRYPT_CTX_DONE = 6
+        }
+
+        private ContextSafeHandle _handle;
+        private Status _status;
+
+        internal CryptContext(ContextSafeHandle handle)
+        {
+            _handle = handle;
+            _status = new Status();
+        }
+
+        /// <summary>
+        /// Gets the state.
+        /// </summary>
+        /// <value>
+        /// The state.
+        /// </value>
+        public StateCode State
+        {
+            get
+            {
+                return Library.mongocrypt_ctx_state(_handle);
+            }
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether this instance is done.
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if this instance is done; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsDone
+        {
+            get { return State == StateCode.MONGOCRYPT_CTX_DONE; }
+        }
+
+        /// <summary>
+        /// Gets the operation.
+        /// </summary>
+        /// <returns>Binary payload to send to either KMS, mongod, or mongocryptd</returns>
+        public Binary GetOperation()
+        {
+            Binary binary = new Binary();
+            Check(Library.mongocrypt_ctx_mongo_op(_handle, binary.Handle));
```

Comment:
Where is the size of the Binary buffer set?

Reply:
In libmongocrypt